### PR TITLE
fix(Reporters): Stryker no longer throws exception on no mutants found

### LINKED ISSUE #1806: Stryker throws for the mutant-free world
**Describe the bug**
Wanted to test the latest feature but got an exception:
```csharp
PS C:\Users\psemkin\Desktop\Repro\Repro.Test> dotnet stryker

   _____ _              _               _   _ ______ _______  
  / ____| |            | |             | \ | |  ____|__   __| 
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |    
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |    
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |    
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|    
                   __/ |                                      
                  |___/                                       


Version: 1.1.0

[14:01:16 INF] Identifying project to mutate.
[14:01:19 INF] The project C:\Users\psemkin\Desktop\Repro\Repro\Repro.csproj will be mutated.
[14:01:21 INF] Analysis complete.
[14:01:21 INF] Building test project C:\Users\psemkin\Desktop\Repro\Repro.Test\Repro.Test.csproj (1/1)
[14:01:28 INF] Total number of tests found: 1.
[14:01:28 INF] Initial testrun started.
[14:01:31 INF] 0 mutants created
[14:01:31 INF] Capture mutant coverage using 'CoverageBasedTest' mode.
[14:01:33 WRN] It's a mutant-free world, nothing to test.
[14:01:34 INF] Time Elapsed 00:00:17.6776698
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at Stryker.Core.Reporters.Progress.ProgressBar.Tick(String message) in /_/src/Stryker.Core/Stryker.Core/Reporters/Progress/ProgressBar.cs:line 36
   at Stryker.Core.Reporters.Progress.ProgressBarReporter.ReportFinalState() in /_/src/Stryker.Core/Stryker.Core/Reporters/Progress/ProgressBarReporter.cs:line 68
   at Stryker.Core.Reporters.Progress.ProgressReporter.OnAllMutantsTested(IReadOnlyProjectComponent reportComponent) in /_/src/Stryker.Core/Stryker.Core/Reporters/Progress/ProgressReporter.cs:line 32
   at Stryker.Core.Reporters.BroadcastReporter.OnAllMutantsTested(IReadOnlyProjectComponent reportComponent) in /_/src/Stryker.Core/Stryker.Core/Reporters/BroadcastReporter.cs:line 63
   at Stryker.Core.StrykerRunner.RunMutationTest(IStrykerInputs inputs, ILoggerFactory loggerFactory, IProjectOrchestrator projectOrchestrator) in /_/src/Stryker.Core/Stryker.Core/StrykerRunner.cs:line 95
   at Stryker.CLI.StrykerCli.RunStryker(IStrykerInputs inputs) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 90
   at Stryker.CLI.StrykerCli.<>c__DisplayClass9_0.<Run>b__0() in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 65
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.<>c__DisplayClass146_0.<OnExecute>b__0(CancellationToken _)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.ExecuteAsync(String[] args, CancellationToken cancellationToken)
   at McMaster.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at Stryker.CLI.StrykerCli.Run(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/StrykerCLI.cs:line 71
   at Stryker.CLI.Program.Main(String[] args) in /_/src/Stryker.CLI/Stryker.CLI/Program.cs:line 15
```

So I guess things work in general, it's just the latest mutation report voodoo broke something.

**Expected behavior**
Generated report with the mutation score 100 - or at least no exception.

**Desktop (please complete the following information):**
 - OS: Windows
 - Type of project: core
 - Framework Version: net6
 - Stryker Version: 1.1.0

**Additional context**
Repro: [Repro.zip](https://github.com/stryker-mutator/stryker-net/files/7546478/Repro.zip)

### PR DESCRIPTION
**Pull request by richardwerkman**
closes #1806

Refactors readonly structure for files and folders

### CODE REVIEW

File: src/Stryker.Core/Stryker.Core/StrykerRunner.cs [hunk 1]
```diff
@@ -156,7 +159,7 @@ private IProjectComponent AddRootFolderIfMultiProject(IEnumerable<IProjectCompon
         {
             if (projectComponents.Count() > 1)
             {
-                var rootComponent = new FolderComposite
+                var rootComponent = new CsharpFolderComposite
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/StrykerRunner.cs:162]: The CsharpFolderComposite is a non-generic FolderComposite that is only valid for csharp projects. This change means that when someone wants to improve on the work that was started for fsharp they will have to reinvent a way to make this generic so the StrykerRunner can work with both csharp and fsharp projects in a generic way.

File: src/Stryker.Core/Stryker.Core/ProjectComponents/ProjectComponent.cs [hunk 2]
```diff
@@ -5,15 +5,36 @@
...
         string RelativePath { get; set; }
+        public Display DisplayFile { get; set; }
+        public Display DisplayFolder { get; set; }
+        IEnumerable<IReadOnlyMutant> TotalMutants();
+        IEnumerable<IReadOnlyMutant> ValidMutants();
+        IEnumerable<IReadOnlyMutant> UndetectedMutants();
+        IEnumerable<IReadOnlyMutant> DetectedMutants();
+        IEnumerable<IReadOnlyMutant> InvalidMutants();
+        IEnumerable<IReadOnlyMutant> IgnoredMutants();
+        IEnumerable<IReadOnlyMutant> NotRunMutants();
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/ProjectComponents/ProjectComponent.cs:24]: These read as properties but are methods, why?

File: src/Stryker.Core/Stryker.Core/ProjectComponents/IFileLeaf.cs [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
         T MutatedSyntaxTree { get; set; }
     }
-    public interface IFileLeaf : IProjectComponent
-    {
-        string SourceCode { get; set; }
 
-        ReadOnlyFileLeaf ToReadOnly();
+    public interface IFileLeaf : IReadOnlyFileLeaf
+    {
+        new string SourceCode { get; set; }
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/ProjectComponents/IFileLeaf.cs:17]: Is the new necessary here if you're only adding a set?

File: src/Stryker.Core/Stryker.Core/ProjectComponents/FolderComposite.cs [hunk 4]
```diff
@@ -21,10 +29,9 @@ public override IEnumerable<Mutant> Mutants
         /// <summary>
         /// Add a syntax tree to this folder that is needed in compilation but should not be mutated
         /// </summary>
-        /// <param name="syntaxTree"></param>
         public void AddCompilationSyntaxTree(T syntaxTree) => _compilationSyntaxTrees.Add(syntaxTree);
-        public override IEnumerable<T> CompilationSyntaxTrees => _compilationSyntaxTrees.Union(
-            Children.Cast<ProjectComponent<T>>().SelectMany(c => c.CompilationSyntaxTrees));
+        public override IEnumerable<T> CompilationSyntaxTrees => _compilationSyntaxTrees.Union(ChildCompilationSyntaxTree);
+        private IEnumerable<T> ChildCompilationSyntaxTree => Children.Cast<ProjectComponent<T>>().SelectMany(c => c.CompilationSyntaxTrees);
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/ProjectComponents/FolderComposite.cs:34]:
```suggestion
        public override IEnumerable<T> CompilationSyntaxTrees => _compilationSyntaxTrees.Union(_childCompilationSyntaxTree);
        private IEnumerable<T> _childCompilationSyntaxTree => Children.Cast<ProjectComponent<T>>().SelectMany(c => c.CompilationSyntaxTrees);
```

File: src/Stryker.Core/Stryker.Core/ProjectComponents/ProjectComponent.cs (hunk 2 above)
Comment (richardwerkman) [reply src/Stryker.Core/Stryker.Core/ProjectComponents/ProjectComponent.cs:24]: Sonar told me to do this. As they iterate lists they should not be properties

File: src/Stryker.Core/Stryker.Core/ProjectComponents/IFileLeaf.cs (hunk 3 above)
Comment (richardwerkman) [reply src/Stryker.Core/Stryker.Core/ProjectComponents/IFileLeaf.cs:17]: Yes, it explicitly overrides the property

**Pull request merged**